parsers/IN_WE.py

```diff
@@ -18,153 +18,144 @@
+API: https://www.wrldc.in/OnlinestateTest1.aspx/GetRealTimeData
+"""
+CONSUMPTION_URL = (
+    f"{IN_WE_PROXY}/OnlinestateTest1.aspx/GetRealTimeData?host=https://www.wrldc.in"
+)
 
-def get_date_range(dt: datetime):
-    return pd.date_range(
-        arrow.get(dt).floor("day").datetime,
-        arrow.get(dt).ceil("day").floor("hour").datetime,
-        freq="H",
-    ).to_pydatetime()
+EXCHANGES_MAPPING = {
+    "IN-SO->IN-WE": "WR-SR",
+    "IN-EA->IN-WE": "WR-ER",
+    "IN-NO->IN-WE": "WR-NR",
+}
 
 
-def fetch_data(
-    kind: str | None = None,
-    session: Session | None = None,
-    target_datetime: datetime | None = None,
-) -> dict:
-    """- get production data from wrldc.in
-    - filter all rows with same hour as target_datetime"""
+def fetch_data(url: str, session: Session, target_datetime: datetime) -> dict:
+    """
+    get data from wrldc.in
+    """
     assert target_datetime is not None
-    assert kind is not None
 
     r = session or Session()
-    payload = {"date": target_datetime.strftime("%Y-%m-%d")}
+    payload = {"date": target_datetime.strftime("%Y-%m-%d"), "Flag": 24}
 
-    resp: Response = r.post(url=KIND_MAPPING[kind]["url"], json=payload)
+    resp: Response = r.post(url=url, json=payload)
 
     try:
         data = json.loads(resp.json().get("d", {}))
     except json.decoder.JSONDecodeError:
         raise ParserException(
             parser="IN_WE.py",
-            message=f"{target_datetime}: {kind} data is not available",
+            message=f"{target_datetime}: {url} data is not available",
         )
-
-    datetime_col = KIND_MAPPING[kind]["datetime_column"]
-    for item in data:
-        item[datetime_col] = datetime.strptime(item[datetime_col], "%Y-%d-%m %H:%M:%S")
-        dt = arrow.get(item[datetime_col])
-        if dt.second >= 30:
-            item[datetime_col] = dt.shift(minutes=1).floor("minute").datetime
-        else:
-            item[datetime_col] = dt.floor("minute").datetime
     return data
 
 
-def filter_raw_data(
-    kind: str,
-    data: dict,
-    target_datetime: datetime,
-) -> pd.DataFrame:
+def process_exchanges_data(
+    data: dict, zone_key1: str, zone_key2: str, logger: Logger
+) -> ExchangeList:
     """
-    Filters out correct datetimes (source data is 12 hour format)
+    Processes exchanges data:
+    - filter out the records that correspond to the zone_key
+    - convert the datetime field to contain only date and hour(remove minute and secods)
+    - set timezone to IST for datetime field
+    - average all data points grouping by target_datetime hour
+    - generate the Exchange list from the Current_Loading column
     """
-    assert len(data) > 0
-    assert kind != ""
-
-    dt_12_hour = arrow.get(target_datetime.strftime("%Y-%m-%d %I:%M")).datetime
-    datetime_col = KIND_MAPPING[kind]["datetime_column"]
-    filtered_data = pd.DataFrame(
-        [item for item in data if item[datetime_col].hour == dt_12_hour.hour]
```

Comment:
There is definitely a bug here. In the older version given the older data had datetime field in 12 hour format with missing AM PM value, this will filter aggregate the data for morning and evening in the same bucket. For example, if the data has records for 2023-11-08 11:00 **AM** and 2023-11-08 11:00 **PM** it will all be considered to be in the same hour

_The older API provided data in a format which can never be correctly interpreted, improving the correctness is the primary motivation I am pushing for this change_

Because of this bug, let's say the parser is run at 2023-11-08 18:23, the source will only provide data till 18:23; however the older version of the parser would generate data in the future as well till 23:00 hour because it will mistakenly copy the data from 11 hour to 23 hour.